### PR TITLE
[install] support git dependencies

### LINKED ISSUE #82: Support git dependencies

### LINKED ISSUE #1703: Can't install git type dependencies
### What version of Bun is running?

0.4.0

### What platform is your computer?

Darwin 22.2.0 arm64 arm

### What steps can reproduce the bug?

Include a dependency with like this one

`"git+https://github.com/ethereumjs/ethereumjs-abi.git"`

### What is the expected behavior?

Install it correctly

### What do you see instead?

Get the following error

`error:  Unsupported dependency type git for "git+https://github.com/ethereumjs/ethereumjs-abi.git"`

### Additional information

_No response_

### LINKED ISSUE #1954: Support git dependencies with access tokens
Bun now supports `github:` dependencies, which also includes `git+https:`, but it does not properly handle access tokens.

`error: Support for dependency type "git" is not implemented yet ("git+https://repo-clone-access:TOKEN@github.com/ORG/repository.git")`

_Originally posted by @dalisoft in https://github.com/oven-sh/bun/issues/1921#issuecomment-1410229681_

### PR DESCRIPTION
**Pull request by alexlamsl**
fixes #82
fixes #1703
fixes #1954

### CODE REVIEW

File: src/install/repository.zig [hunk 1]
```diff
@@ -87,4 +99,164 @@ pub const Repository = extern struct {
             }
         }
     };
+
+    fn exec(allocator: std.mem.Allocator, env: *DotEnv.Loader, cwd_dir: std.fs.Dir, argv: []const string) !string {
+        const buf_map = try env.map.cloneToEnvMap(allocator);
+        const result = try std.ChildProcess.exec(.{
```
Comment (Jarred-Sumner) [inline src/install/repository.zig:105]: There's a way we can do this same thread non-blocking but it's more complicated

File: src/install/repository.zig [hunk 2]
```diff
@@ -87,4 +99,164 @@ pub const Repository = extern struct {
...
+
+            _ = exec(allocator, env, cache_dir, &[_]string{
+                "git",
+                "clone",
+                "--quiet",
+                "--no-checkout",
+                try bun.getFdPath(repo_dir.fd, &final_path_buf),
+                folder_name,
+            }) catch |err| {
+                Output.prettyErrorln("<r><red>Error<r> \"git clone\" failed for {s}", .{
```
Comment (Jarred-Sumner) [inline src/install/repository.zig:199]: These need to use logger.Log I think, to handle silent mode

File: src/install/install.zig [hunk 3]
```diff
@@ -487,30 +488,41 @@ const Task = struct {
...
+            hasher.update(url);
+            return @as(u64, 4 << 61) | @as(u64, @truncate(u61, hasher.final()));
+        }
+
+        pub fn forGitCheckout(url: string, resolved: string) u64 {
+            var hasher = std.hash.Wyhash.init(0);
+            hasher.update(url);
+            hasher.update("@");
+            hasher.update(resolved);
+            return @as(u64, 5 << 61) | @as(u64, @truncate(u61, hasher.final()));
```
Comment (Jarred-Sumner) [inline src/install/install.zig:525]: 5 + 61 == 66?

I think we need to make the right hand side u59

File: src/install/repository.zig (hunk 1 above)
Comment (alexlamsl) [reply src/install/repository.zig:105]: I explored that a little &minus; basically `if (child_process.term) |term| ...` in some form of loop &minus; but can't convince myself the benefit was worth the trouble :sweat_smile:

File: src/install/install.zig (hunk 3 above)
Comment (alexlamsl) [reply src/install/install.zig:525]: `64 - 61 = 3` bits, and `2 ** 3 = 8` so we have `0 .. 7` to play with here? :thinking:

Comment (Jarred-Sumner) [reply src/install/install.zig:525]: ah yeah

File: src/install/repository.zig (hunk 1 above)
Comment (Jarred-Sumner) [reply src/install/repository.zig:105]: we'd use Subprocess from Bun.spawn to do that instead of going through std.ChildProcess

**Comment by Jarred-Sumner** [comment]
Thank you

**Comment by razvanphp** [comment]
I think npm had a short form that was working too:
```
  "dependencies": {
    "simple-peer": "github:vampirebyte/simple-peer#next",
```
I had to prepend `git+https://git@github.com/` instead.

**Comment by rawpixel-vincent** [comment]
> I think npm had a short form that was working too:
> 
> ```
>   "dependencies": {
>     "simple-peer": "github:vampirebyte/simple-peer#next",
> ```
> 
> I had to prepend `git+https://git@github.com/` instead.

thanks that worked for me

after upgrading from 1.1.29, `git+https://github.com/...` git dependencies stopped working.

using `git+https://git@github.com/...` works

**Pull request merged**